cpp/src/file/tsfile_io_writer.cc

```diff
@@ -405,30 +417,67 @@ int TsFileIOWriter::write_file_index() {
     }
 
     if (IS_SUCC(ret)) {
-        MetaIndexNode *device_index_root_node = nullptr;
-        if (RET_FAIL(build_device_level(device_map, device_index_root_node,
-                                        writing_mm))) {
+        if (!generate_table_schema_) {
+            MetaIndexNode *device_index_root_node = nullptr;
+            if (RET_FAIL(build_device_level(device_map, device_index_root_node,
+                                            writing_mm))) {
+            } else {
+                TsFileMeta tsfile_meta;
+                tsfile_meta.index_node_ = device_index_root_node;
+                tsfile_meta.meta_offset_ = meta_offset;
+                tsfile_meta.bloom_filter_ = &filter;
+                int64_t tsfile_meta_offset = cur_file_position();
+                uint32_t size = 0;  // cppcheck-suppress unreadVariable
+                OFFSET_DEBUG("before tsfile_meta written");
+                if (RET_FAIL(tsfile_meta.serialize_to(write_stream_))) {
+                } else if (RET_FAIL(filter.serialize_to(write_stream_))) {
+                } else {
+                    int64_t tsfile_meta_end_offset = cur_file_position();
+                    size =
+                        (uint32_t)(tsfile_meta_end_offset - tsfile_meta_offset);
+                    ret = SerializationUtil::write_ui32(size, write_stream_);
+                }
+#if DEBUG_SE
+                std::cout << "writer tsfile_meta: " << tsfile_meta
+                          << ", tsfile_meta_offset=" << tsfile_meta_offset
+                          << ", size=" << size << std::endl;
+#endif
+                tsfile_meta.index_node_ =
+                    nullptr;  // memory management is delegated to writing_mm
+                tsfile_meta.bloom_filter_ = nullptr;
+            }
         } else {
```

Comment:
I do not think `generate_table_schema_` should control this. There are no two ways of constructing Metadata, only one way.